# Practice Programming Assignment: swirl Lesson 2: vapply and tapply

In the last lesson, you learned about the two most fundamental members of R's *apply family of functions: lapply() and sapply(). Both take a list as input, apply a function to each element of the list, then combine and return the result. lapply() always returns a list, whereas sapply() attempts to simplify the result.

In this lesson, you'll learn how to use vapply() and tapply(), each of which serves a very specific purpose within the Split-Apply-Combine methodology. For consistency, we'll use the same dataset we used in the 'lapply and sapply' lesson.

The Flags dataset from the UCI Machine Learning Repository contains details of various nations and their flags. More information may be found here: http://archive.ics.uci.edu/ml/datasets/Flags

We've stored the data in a variable called flags. If it's been a while since you completed the 'lapply and sapply' lesson, you may want to reacquaint yourself with the data by using functions like dim(), head(), str(), and summary() when you return to the prompt (>). You can also type viewinfo() at the prompt to bring up some documentation for the dataset. Let's get started!

As you saw in the last lesson, the unique() function returns a vector of the unique values contained in the object passed to it. Therefore, sapply(flags, unique) returns a list containing one vector of unique values for each column of the flags dataset. Try it again now.

In [1]:
# Load dataset
flags <- read.csv("./data/flag.data.txt", header=FALSE)
# Set column names
colnames(flags) <- c("name", "landmass", "zone", "area", "population",
                     "language", "religion", "bars", "stripes", "colours",
                     "red", "green", "blue", "gold", "white", "black",
                     "orange", "mainhue", "circles", "crosses", "saltires",
                     "quarters", "sunstars", "crescent", "triangle",
                     "icon", "animate", "text", "topleft", "botright")

In [2]:
sapply(flags, unique)

$name
  [1] "Afghanistan"              "Albania"                 
  [3] "Algeria"                  "American-Samoa"          
  [5] "Andorra"                  "Angola"                  
  [7] "Anguilla"                 "Antigua-Barbuda"         
  [9] "Argentina"                "Argentine"               
 [11] "Australia"                "Austria"                 
 [13] "Bahamas"                  "Bahrain"                 
 [15] "Bangladesh"               "Barbados"                
 [17] "Belgium"                  "Belize"                  
 [19] "Benin"                    "Bermuda"                 
 [21] "Bhutan"                   "Bolivia"                 
 [23] "Botswana"                 "Brazil"                  
 [25] "British-Virgin-Isles"     "Brunei"                  
 [27] "Bulgaria"                 "Burkina"                 
 [29] "Burma"                    "Burundi"                 
 [31] "Cameroon"                 "Canada"                  
 [33] "Cape-Verde-Islands"       "Cayman-Islands"          
 [35] "Central-African-Republic" "Chad"                    
 [37] "Chile"                    "China"                   
 [39] "Colombia"                 "Comorro-Islands"         
 [41] "Congo"                    "Cook-Islands"            
 [43] "Costa-Rica"               "Cuba"                    
 [45] "Cyprus"                   "Czechoslovakia"          
 [47] "Denmark"                  "Djibouti"                
 [49] "Dominica"                 "Dominican-Republic"      
 [51] "Ecuador"                  "Egypt"                   
 [53] "El-Salvador"              "Equatorial-Guinea"       
 [55] "Ethiopia"                 "Faeroes"                 
 [57] "Falklands-Malvinas"       "Fiji"                    
 [59] "Finland"                  "France"                  
 [61] "French-Guiana"            "French-Polynesia"        
 [63] "Gabon"                    "Gambia"                  
 [65] "Germany-DDR"              "Germany-FRG"             
 [67] "Ghana"                    "Gibraltar"               
 [69] "Greece"                   "Greenland"               
 [71] "Grenada"                  "Guam"                    
 [73] "Guatemala"                "Guinea"                  
 [75] "Guinea-Bissau"            "Guyana"                  
 [77] "Haiti"                    "Honduras"                
 [79] "Hong-Kong"                "Hungary"                 
 [81] "Iceland"                  "India"                   
 [83] "Indonesia"                "Iran"                    
 [85] "Iraq"                     "Ireland"                 
 [87] "Israel"                   "Italy"                   
 [89] "Ivory-Coast"              "Jamaica"                 
 [91] "Japan"                    "Jordan"                  
 [93] "Kampuchea"                "Kenya"                   
 [95] "Kiribati"                 "Kuwait"                  
 [97] "Laos"                     "Lebanon"                 
 [99] "Lesotho"                  "Liberia"                 
[101] "Libya"                    "Liechtenstein"           
[103] "Luxembourg"               "Malagasy"                
[105] "Malawi"                   "Malaysia"                
[107] "Maldive-Islands"          "Mali"                    
[109] "Malta"                    "Marianas"                
[111] "Mauritania"               "Mauritius"               
[113] "Mexico"                   "Micronesia"              
[115] "Monaco"                   "Mongolia"                
[117] "Montserrat"               "Morocco"                 
[119] "Mozambique"               "Nauru"                   
[121] "Nepal"                    "Netherlands"             
[123] "Netherlands-Antilles"     "New-Zealand"             
[125] "Nicaragua"                "Niger"                   
[127] "Nigeria"                  "Niue"                    
[129] "North-Korea"              "North-Yemen"             
[131] "Norway"                   "Oman"                    
[133] "Pakistan"                 "

What if you had forgotten how unique() works and mistakenly thought it returns the *number* of unique values contained in the object passed to it? Then you might have incorrectly expected sapply(flags, unique) to return a numeric vector, since each element of the list returned would contain a single number and sapply() could then simplify the result to a vector.

When working interactively (at the prompt), this is not much of a problem, since you see the result immediately and will quickly recognize your mistake. However, when working non-interactively (e.g. writing your own functions), a misunderstanding may go undetected and cause incorrect results later on. Therefore, you may wish to be more careful and that's where vapply() is useful.

Whereas sapply() tries to 'guess' the correct format of the result, vapply() allows you to specify it explicitly. If the result doesn't match the format you specify, vapply() will throw an error, causing the operation to stop. This can prevent significant problems in your code that might be caused by getting unexpected return values from sapply().

Try vapply(flags, unique, numeric(1)), which says that you expect each element of the result to be a numeric vector of length 1. Since this is NOT actually the case, YOU WILL GET AN ERROR.

In [3]:
vapply(flags, unique, numeric(1))

ERROR: Error in vapply(flags, unique, numeric(1)): values must be length 1,
 but FUN(X[[1]]) result is length 194


Recall from the previous lesson that sapply(flags, class) will return a character vector containing the class of each column in the dataset. Try that again now to see the result.

In [4]:
sapply(flags, class)

name    landmass        zone        area  population    language 
"character"   "integer"   "integer"   "integer"   "integer"   "integer" 
   religion        bars     stripes     colours         red       green 
  "integer"   "integer"   "integer"   "integer"   "integer"   "integer" 
       blue        gold       white       black      orange     mainhue 
  "integer"   "integer"   "integer"   "integer"   "integer" "character" 
    circles     crosses    saltires    quarters    sunstars    crescent 
  "integer"   "integer"   "integer"   "integer"   "integer"   "integer" 
   triangle        icon     animate        text     topleft    botright 
  "integer"   "integer"   "integer"   "integer" "character" "character"

If we wish to be explicit about the format of the result we expect, we can use vapply(flags, class, character(1)). The 'character(1)' argument tells R that we expect the class function to return a character vector of length 1 when applied to EACH column of the flags dataset. Try it now.

In [5]:
vapply(flags, class, character(1))

name    landmass        zone        area  population    language 
"character"   "integer"   "integer"   "integer"   "integer"   "integer" 
   religion        bars     stripes     colours         red       green 
  "integer"   "integer"   "integer"   "integer"   "integer"   "integer" 
       blue        gold       white       black      orange     mainhue 
  "integer"   "integer"   "integer"   "integer"   "integer" "character" 
    circles     crosses    saltires    quarters    sunstars    crescent 
  "integer"   "integer"   "integer"   "integer"   "integer"   "integer" 
   triangle        icon     animate        text     topleft    botright 
  "integer"   "integer"   "integer"   "integer" "character" "character"

Note that since our expectation was correct (i.e. character(1)), the vapply() result is identical to the sapply() result -- a character vector of column classes.

You might think of vapply() as being 'safer' than sapply(), since it requires you to specify the format of the output in advance, instead of just allowing R to 'guess' what you wanted. In addition, vapply() may perform faster than sapply() for large datasets. However, when doing data analysis interactively (at the prompt), sapply() saves you some typing and will often be good enough.

As a data analyst, you'll often wish to split your data up into groups based on the value of some variable, then apply a function to the members of each group. The next function we'll look at, tapply(), does exactly that.

Use ?tapply to pull up the documentation.

In [6]:
?tapply

tapply                  package:base                   R Documentation

_A_p_p_l_y _a _F_u_n_c_t_i_o_n _O_v_e_r _a _R_a_g_g_e_d _A_r_r_a_y

_D_e_s_c_r_i_p_t_i_o_n:

     Apply a function to each cell of a ragged array, that is to each
     (non-empty) group of values given by a unique combination of the
     levels of certain factors.

_U_s_a_g_e:

     tapply(X, INDEX, FUN = NULL, ..., default = NA, simplify = TRUE)
     
_A_r_g_u_m_e_n_t_s:

       X: an R object for which a ‘split’ method exists.  Typically
          vector-like, allowing subsetting with ‘[’.

   INDEX: a ‘list’ of one or more ‘factor’s, each of same length as
          ‘X’.  The elements are coerced to factors by ‘as.factor’.

     FUN: a function (or name of a function) to be applied, or ‘NULL’.
          In the case of functions like ‘+’, ‘%*%’, etc., the function
          name must be backquoted or quoted.  If ‘FUN’ is ‘NULL’,
          tapply returns a vec

The 'landmass' variable in our dataset takes on integer values between 1 and 6, each of which represents a different part of the world. Use table(flags$landmass) to see how many flags/countries fall into each group.

In [7]:
table(flags$landmass)


 1  2  3  4  5  6 
31 17 35 52 39 20 

The 'animate' variable in our dataset takes the value 1 if a country's flag contains an animate image (e.g. an eagle, a tree, a human hand) and 0 otherwise. Use table(flags$animate) to see how many flags contain an animate image.

In [8]:
table(flags$animate)


  0   1 
155  39 

This tells us that 39 flags contain an animate object (animate = 1) and 155 do not (animate = 0).

If you take the arithmetic mean of a bunch of 0s and 1s, you get the proportion of 1s. Use tapply(flags$animate, flags$landmass, mean) to apply the mean function to the 'animate' variable separately for each of the six landmass groups, thus giving us the proportion of flags containing an animate image WITHIN each landmass group.

In [9]:
tapply(flags$animate, flags$landmass, mean)

1         2         3         4         5         6 
0.4193548 0.1764706 0.1142857 0.1346154 0.1538462 0.3000000

The first landmass group (landmass = 1) corresponds to North America and contains the highest proportion of flags with an animate image (0.4194).

Similarly, we can look at a summary of population values (in round millions) for countries with and without the color red on their flag with tapply(flags$population, flags$red, summary).

In [10]:
tapply(flags$population, flags$red, summary)

$`0`
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   0.00    0.00    3.00   27.63    9.00  684.00 

$`1`
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
    0.0     0.0     4.0    22.1    15.0  1008.0 


Lastly, use the same approach to look at a summary of population values for each of the six landmasses.

In [11]:
tapply(flags$population, flags$landmass, summary)

$`1`
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   0.00    0.00    0.00   12.29    4.50  231.00 

$`2`
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   0.00    1.00    6.00   15.71   15.00  119.00 

$`3`
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   0.00    0.00    8.00   13.86   16.00   61.00 

$`4`
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  0.000   1.000   5.000   8.788   9.750  56.000 

$`5`
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   0.00    2.00   10.00   69.18   39.00 1008.00 

$`6`
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   0.00    0.00    0.00   11.30    1.25  157.00 


In this lesson, you learned how to use vapply() as a safer alternative to sapply(), which is most helpful when writing your own functions. You also learned how to use tapply() to split your data into groups based on the value of some variable, then apply a function to each group. These functions will come in handy on your quest to become a better data analyst.